src/frequenz/channels/_latest_value_cache.py

```diff
@@ -48,15 +57,56 @@ def __str__(self) -> str:
         return "<no value received yet>"
 
 
-class LatestValueCache(typing.Generic[T_co]):
+class LatestValueCache(typing.Generic[T_co, HashableT]):
     """A cache that stores the latest value in a receiver.
 
     It provides a way to look up the latest value in a stream without any delay,
     as long as there has been one value received.
     """
 
+    @typing.overload
+    def __init__(
+        self: "LatestValueCache[T_co, None]",
+        receiver: Receiver[T_co],
+        *,
+        unique_id: str | None = None,
+        key: None = None,
+    ) -> None:
+        """Create a new cache that does not use keys.
+
+        Args:
+            receiver: The receiver to cache.
+            unique_id: A string to help uniquely identify this instance. If not
+                provided, a unique identifier will be generated from the object's
+                [`id()`][id]. It is used mostly for debugging purposes.
+            key: This parameter is ignored when set to `None`.
+        """
+
+    @typing.overload
+    def __init__(
+        self: "LatestValueCache[T_co, HashableT]",
+        receiver: Receiver[T_co],
+        *,
+        unique_id: str | None = None,
+        key: typing.Callable[[T_co], HashableT],
+    ) -> None:
```

Comment:
Damn, is this overload over `self` really necessary? Doesn't `self` just works magically if you don't annotate it at all? Have you tried with `Self`?

Also, you should `from __future__ import annotations` and remove the quotes from the types if you still need it.

Reply:
I wonder if using a (non-hashable) sentinel in the type could help with this too, if the issue is that `None` is `Hashable`.